---
title: "Exercises part 1"
author: "<YOUR NAME>"
date: "<YOUR DATE>"
output:
  html_document:
    df_print: paged
subtitle: 'Course: Data analysis and visualization using R'
---

This R Markdown document contains exercises to accompany the course "Data analysis and visualization using R".  
This document contains the exercises themselves plus (in most cases) an R code chunk to fill, complete, or correct.
For more details on using R Markdown see <http://rmarkdown.rstudio.com>.  

Some assignments have test code to verify your solution. They are marked with   
`###### TEST CODE - DO NOT REMOVE OR EDIT #######`  

----


```{r setup, echo = FALSE, include = FALSE}
## THIS IS SETUP CODE - DO NOT EDIT!!!                    ##
labelerURL <- "https://raw.githubusercontent.com/MichielNoback/davur1/gh-pages/exercises/labeler.R"
source(labelerURL)
labelMaker <- labeler() 
```


# `r labelMaker$getSection()` Basic plotting

With all plots, take care to adhere to the rules regarding titles and other decorations. Tip: the site [Quick-R](http://www.statmethods.net/) has nice detailed information with examples on the different plot types and their configuration. Especially the section [on plotting](http://www.statmethods.net/graphs/index.html) is helpful for these assignments.


#### **`r labelMaker$getExercise()` ** 

The vectors below hold data for a staircase walking experiment. A subject of normal weight and height was asked to ascend a (long) stairs wearing a heart-rate monitor. The subjects' heart was registered for different step heights. Create a **line plot** showing the dependence of heart rate (y axis) on stair height (x axis).


```{r fig.width = 5, fig.height = 3}
#number of steps on the stairs
stair_height <- c(0, 5, 10, 15, 20, 25, 30, 35)
#heart rate after ascending the stairs
heart_rate <- c(66, 65, 67, 69, 73, 79, 86, 97)
##your code here creating the plot
```

```{r}
#<YOUR CODE HERE>
```


#### **`r labelMaker$getExercise()` ** 
The experiment from the previous question was extended with three more subjects. One of these subjects was like the first of normal weight, whereas the two others were obese. The data are given below. Create a single **scatter plot** with connector lines between the points showing the data for all four subjects. Give the normal-weighted subjects a green line and symbol and the obese subjects a red line and symbol.  
You can add new data series to a plot by using the `points(x, y)` function. Use the `ylim()` function to adjust the Y-axis range.


```{r fig.width = 4.5, fig.height = 3.5}
#number of steps on the stairs
stair_height <- c(0, 5, 10, 15, 20, 25, 30, 35)
#heart rates for subjects with normal weight
heart_rate_1 <- c(66, 65, 67, 69, 73, 79, 86, 97)
heart_rate_2 <- c(61, 61, 63, 68, 74, 81, 89, 104)
#heart rates for obese subjects
heart_rate_3 <- c(58, 60, 67, 71, 78, 89, 104, 121)
heart_rate_4 <- c(69, 73, 77, 83, 88, 96, 102, 127)
```


```{r}
#<YOUR CODE HERE>
```


#### **`r labelMaker$getExercise()` ** 
The body weights of chicks were measured at birth and every second day thereafter until day 20. They were also measured on day 21. In the experiment there were four groups of chicks on different protein diets. Here are the data for the first four chicks. Chick one and two were on diet 1 and chick three and four were on diet 2. Create a single line plot showing the data for all four chicks. Give each chick its own color


```{r fig.width = 4.5, fig.height = 3.5}
# chick weight data
time <- c(0, 2, 4, 6, 8, 10, 12, 14, 16, 18, 20, 21)
chick_1 <- c(42, 51, 59, 64, 76, 93, 106, 125, 149, 171, 199, 205)
chick_2 <- c(40, 49, 58, 72, 84, 103, 122, 138, 162, 187, 209, 215)
chick_3 <- c(42, 53, 62, 73, 85, 102, 123, 138, 170, 204, 235, 256)
chick_4 <- c(41, 49, 61, 74, 98, 109, 128, 154, 192, 232, 280, 290)
```

```{r}
#<YOUR CODE HERE>
```


#### **`r labelMaker$getExercise()` ** 
With the data from the previous question, create a barplot of the maximum weights of the chicks.


```{r fig.width = 4.5, fig.height = 3.5}
#<YOUR CODE HERE>
```


#### **`r labelMaker$getExercise()` ** 
 The R language comes with a wealth of datasets for you to use as practice materials. We will see several of these. One of these datasets is The Time-Series dataset called `discoveries` holding the numbers of "great" inventions and scientific discoveries in each year from 1860 to 1959. Type its name in the console to see it. Create plot(s) answering these two questions:  
 
**`r labelMaker$getLetter(reset = T)`** What is the frequency distribution of numbers of discoveries per year? Use the `hist()` function for this.  

**`r labelMaker$getLetter()`** What is the 5-number summary of discoveries per year?  

**`r labelMaker$getLetter()`** What is the trend over time for the numbers of discoveries per year?

PS actually this is not a simple vector, but a vector with some time=-related attributes called a Time-Series (a `ts` class), but this does not really matter for this assignment.

```{r fig.width = 4.5, fig.height = 3.5}
#<YOUR CODE HERE>
```


#### **`r labelMaker$getExercise()` **  

The R datasets package has three related timeseries datasets relating to lung cancer deaths. These are `ldeaths`, `mdeaths` and `fdeaths` for total, male and female deatchs, respectively. Create a line plot showing the montly mortality holding all three of these datasets. Use the `legend()` function to add a legend to the plot, as shown in this example:


```{r fig.width = 4.5, fig.height = 3.5}
t <- 1:5
y1 <- c(2, 3, 5, 4, 6)
y2 <- c(1, 3, 4, 5, 7)
plot(t, y1, type = "b", ylab = "response", ylim = c(0, 8))
points(t, y2, col = "blue", type = "b")
legend("topleft", legend = c("series 1", "series 2"), col = c("black", "blue"), pch = 1, lty = 1)
```

**`r labelMaker$getLetter(reset = T)`** Create the mentioned line plot. Do you see trends and/or patterns and if so, can you explain these?

**`r labelMaker$getLetter()`** Create a combined boxplot of the three time-series. Are there outliers? If so, can you figure out when this occurred?

```{r fig.width = 5.5, fig.height = 4.5}
#<YOUR CODE HERE>
```


# `r labelMaker$getSection()` Complex datatypes & Basic (dataframe) functions

This section serves you some datatype challenges.


#### **`r labelMaker$getExercise()` ** 

Almost all programming languages know the (hash)map / dictionary data structure storing ao called "key-value" pairs. R does not have a dictionary type but you could make a dict-like structure using a list with named elements. Here is an example.

If I wanted to create and use a DNA codon translation table, and use it to translate a piece of DNA, I could do something like what is shown below (there are only 4 of the 64 codons). See if you can figure out what is going on there


```{r}
## define codon table as list
codons <- list(GGA = "Gly", CCU = "Pro", AAA = "Lys", AGU = "Ser")
## the DNA to translate
my_DNA <- "GGACCUAAAAGU"
my_prot <- ""
## iterate the DNA and take only every position
for (i in seq(1, nchar(my_DNA), by=3)) {
    codon <- substr(my_DNA, i, i+2);
    my_prot <- paste(my_prot, codons[[codon]])
}
print(my_prot)
```

**`r labelMaker$getLetter(reset = T)`** 
Make a modified copy of this code chunk in such a way that no spaces are present between the amino acid residues (use help on `paste()` to figure this out) and that single-letter codes of amino acids are used instead of three-letter codes.


```{r}
#<YOUR CODE HERE>
```

**`r labelMaker$getLetter()`** 
**[Challenge]**. Here is a list called `nuc.weights` with named elements containing two vectors - a vector with the Nucleotide single letter codes and a vector with their molecular weights. Iterate `my.DNA` and calculate its molecular weight using this data structure.


```{r}
nuc_weights <- list(
    nucs = c("A", "C", "G", "U"), 
    weights = c(491.2, 467.2, 507.2, 482.2))
```

```{r}
#<YOUR CODE HERE>
```


#### **`r labelMaker$getExercise()` ** 

The `airquality` dataset is also one of the datasets included in the `datasets` package. We'll explore this for a few questions.

**`r labelMaker$getLetter()`** 
Create a scatterplot of Temperature as a function of Solar radiation. Is there, as you might naively expect, a strong correlation?

```{r fig.width = 4.5, fig.height = 3}
#<YOUR CODE HERE>
```

**`r labelMaker$getLetter(reset = T)`** 
Create a boxplot-series of Temp as a function of Month (use `?boxplot` to find out how this works). What appears to be the warmest month?

```{r fig.width = 4.5, fig.height = 3}
#<YOUR CODE HERE>
```

**`r labelMaker$getLetter()`** 
What date (day/month) has the lowest recorded temperature? Which the highest? Please give temperature values in Celcius, not Fahrenheit! (Yes, this is an extra challenge!)

```{r}
#<YOUR CODE HERE>
```


**`r labelMaker$getLetter()`** 
Create a histogram of the wind speeds, and add a thick blue vertical line for the value of the mean and a fat red line for the median (use `abline()` for this).

```{r fig.width = 4.5, fig.height = 3}
#<YOUR CODE HERE>
```


**`r labelMaker$getLetter()`** 
Use the `pairs()` function with argument `panel = panel.smooth` to plot all pairwise correlations between Ozon, Solar radiation, Wind and Temperature. Which pair shows the strongest correlation?

```{r}
#<YOUR CODE HERE>
```


#### **`r labelMaker$getExercise()` ** 

You will explore a bird observation dataset, downloaded from [GOLDEN GATE AUDUBON SOCIETY](http://goldengateaudubon.org/birding-resources/observations/). This file lists bird observations collected by this bird monitoring group in the San Fransisco Bay Area. I already cleaned it a bit and placed it [here](data/Observations-Data-2014.csv) 

Load the observation data into R and assign it to a variable called `bird_obs`.

```{r eval=TRUE}
#<YOUR CODE HERE>
```

From here on, it is assumed that you have the dataframe `bird_obs` loaded. This series of exercises deals with cleaning and transforming data, and exploring a cleaned dataset using basic plotting techniques and descriptive statistics.


**`r labelMaker$getLetter(reset = T)`** 
First, explore the raw data as they are. 

- What data on bird observations were recorded (i.e. what kind of variables do you have)?
- What did R do to the original column names?
- Are all column names clear to you?

```{r}
#<YOUR CODE HERE>
```

Apparantly, all is loaded as a factor; also the Date.start, Date.end (should be dates of course), and Number (should be Integer) and Notes (should be Character) columns. In the original column names tehre are spaces and these are replaced by dots. First column `Species..` is a serial number and the second `Species` is the English species name.
 
**`r labelMaker$getLetter()`** 
How many bird observations were recorded?

```{r}
#<YOUR CODE HERE>
```

**`r labelMaker$getLetter()`** 
The column holding observation "Number" is actually not a number. Into what type has R converted it?

```{r}
#<YOUR CODE HERE>
```

**`r labelMaker$getLetter()`** 
Convert the "Number" column into an integer column using `as.integer()`, but assign it to a new column called "Count" (i.e. do not overwrite the original values). Compare the first 50 values or so of these two columns. What happened to the data? Is this OK?

```{r}
#<YOUR CODE HERE>
```


**`r labelMaker$getLetter()`** 
The previous question has shown that converting factors to numbers is a bit dangerous. It is often easiest to convert characters to numbers. The best way to do this is by using the `as.is = c(<column indices>)` argument for the `read.table()` function. 

So, which columns should be loaded as real factor data and which as plain character data? Use `read.table()` and the `as.is` argument to reload the data, and then transform the `Number` column to integer again. 

```{r}
#<YOUR CODE HERE>
```

**`r labelMaker$getLetter()`** 
Compare the first 50 values of the Number and Count columns again. Has the conversion succeeded? How many `Number` values could not be transformed into an integer value? Hint: use `is.na()`

```{r}
#<YOUR CODE HERE>
```

**`r labelMaker$getLetter()`** 
Explore the sighting counts:

- What is the maximum number of birds in a single sighting? (Use max() and which() or is.na() to solve this)
- What is the mean sighting count
- What is the median of the sighting count

```{r}
#<YOUR CODE HERE>
```

Is the `Count` variable a normal distributed value? You can use `table(...)` of `plot(density(...))` to explore this further.

```{r fig.width = 4.5, fig.height = 3}
#<YOUR CODE HERE>
```


**`r labelMaker$getLetter()`** 
Explore the species constitution: 

- How many different species were recorded?
- How many genera do they constitute?
- What species from the genus "Puffinus" have been observed?

Hint: use the function `unique()` here.

```{r}
#<YOUR CODE HERE>
```


**`r labelMaker$getLetter()`** 
This is a challenge exercise for those who like to grind their brains! Think of a strategy to "rescue" the NAs that appear after transforming "Number" to "Count". Hint: use `gsub()` or`grep()`

```{r}
#<YOUR CODE HERE>
```


# `r labelMaker$getSection()` Flow control & Functions in R

This section serves you some exercises that will help you improve your function-writing skills.


## Illegal reproductions
As an exercise, you will re-invent the wheel here for some statistical functions.

#### **`r labelMaker$getExercise()` ** 
Create a function, `my_mean()`, that duplicates the R function `mean()`, i.e. calculates and returns the mean of a vector of numbers, without actually using `mean()`. 

```{r}
#<YOUR CODE HERE>
```


Run the chunk below to see if your solution is correct.


```{r echo=FALSE, eval=FALSE}
###### TEST CODE - DO NOT REMOVE OR EDIT #######
x <- c(2, 3, 4, 5, 6)
y <- c(3.5, 2.3, 6.1, 2.2, 3.7, 5.1)
if (my_mean(x) == mean(x) & my_mean(y) == mean(y)) {
    print("You rock!")
} else {
    warning("your implementation of the 'MyMean' function is incorrect")
}
###### TEST CODE - DO NOT REMOVE OR EDIT #######
```



#### **`r labelMaker$getExercise()` ** 
Create a function, `my_sd()`, that duplicates the R function `sd()`, i.e. calculates and returns the standard deviation of a vector of numbers, without actually using `sd()`. 

```{r}
#<YOUR CODE HERE>
```



```{r echo=FALSE, eval=FALSE}
###### TEST CODE - DO NOT REMOVE OR EDIT #######
if (my_sd(x) == sd(x) & my_sd(y) == sd(y)) {
    print("You rock!")
} else {
    warning("your implementation of the 'my_sd' function is incorrect")
}
###### TEST CODE - DO NOT REMOVE OR EDIT #######
```


#### **`r labelMaker$getExercise()` ** 
Create a function, `my_median()`, that duplicates the R function `median()`, i.e. calculates and returns the median of a vector of numbers. This is actually a bit harder than you might expect. Hint: use the `sort()` function.

```{r}
#<YOUR CODE HERE>
```



```{r echo=FALSE, eval=FALSE}
###### TEST CODE - DO NOT REMOVE OR EDIT #######
if (my_median(x) == median(x) & my_median(y) == median(y)) {
    print("You rock!")
} else {
    warning("your implementation of the 'my_median' function is incorrect")
}
###### TEST CODE - DO NOT REMOVE OR EDIT #######
```


#### **`r labelMaker$getExercise()` **  

**[Challenge]** Create a function, `GC_perc()`, that calculates and returns the GC percentage of a DNA or RNA sequence. Accept as input a sequence and a flag -`strict`- indicating whether other characters are accepted than core DNA/RNA (GATUC). If `strict = FALSE`, the percentage of other characters should be reported using a `warning()` call. If `strict = TRUE`, the function should terminate with an error message. Use `stop()` for this. `strict` should default to `TRUE`. NOTE, usage of `strict` can complicate things, so start with the core functionality!
You can use `strsplit()` or `substr()` to get hold of individual sequence characters.

```{r}
#<YOUR CODE HERE>
```


## END OF PART ONE





