<!DOCTYPE html>
<html >

<head>

  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <title>Data Analysis and Visualization using R</title>
  <meta name="description" content="Data Analysis and Visualization using R">
  <meta name="generator" content="bookdown 0.7 and GitBook 2.6.7">

  <meta property="og:title" content="Data Analysis and Visualization using R" />
  <meta property="og:type" content="book" />
  
  
  
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Data Analysis and Visualization using R" />
  
  
  

<meta name="author" content="Michiel Noback, PhD">



  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black">
  
  
<link rel="prev" href="apply.html">
<link rel="next" href="ttest.html">
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />







<script src="libs/htmlwidgets-1.0/htmlwidgets.js"></script>
<link href="libs/datatables-css-0.0.0/datatables-crosstalk.css" rel="stylesheet" />
<script src="libs/datatables-binding-0.4/datatables.js"></script>
<link href="libs/dt-core-1.10.16/css/jquery.dataTables.min.css" rel="stylesheet" />
<link href="libs/dt-core-1.10.16/css/jquery.dataTables.extra.css" rel="stylesheet" />
<script src="libs/dt-core-1.10.16/js/jquery.dataTables.min.js"></script>
<link href="libs/crosstalk-1.0.0/css/crosstalk.css" rel="stylesheet" />
<script src="libs/crosstalk-1.0.0/js/crosstalk.min.js"></script>


<style type="text/css">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; position: absolute; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; }
pre.numberSource a.sourceLine:empty
  { position: absolute; }
pre.numberSource a.sourceLine::before
  { content: attr(data-line-number);
    position: absolute; left: -5em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">DAVuR Course Syllabus</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Resources, Exercises and Presentations</a></li>
<li class="chapter" data-level="2" data-path="usecases.html"><a href="usecases.html"><i class="fa fa-check"></i><b>2</b> Example Use Cases</a><ul>
<li class="chapter" data-level="2.1" data-path="dfselection.html"><a href="dfselection.html"><i class="fa fa-check"></i><b>2.1</b> Dataframe Selections</a></li>
<li class="chapter" data-level="2.2" data-path="apply.html"><a href="apply.html"><i class="fa fa-check"></i><b>2.2</b> Apply</a></li>
<li class="chapter" data-level="2.3" data-path="embeddeddf.html"><a href="embeddeddf.html"><i class="fa fa-check"></i><b>2.3</b> Processing Embedded Dataframes</a></li>
<li class="chapter" data-level="2.4" data-path="ttest.html"><a href="ttest.html"><i class="fa fa-check"></i><b>2.4</b> Applying statistical tests on Dataframes</a><ul>
<li class="chapter" data-level="2.4.1" data-path="ttest.html"><a href="ttest.html#intro"><i class="fa fa-check"></i><b>2.4.1</b> Intro</a></li>
<li class="chapter" data-level="2.4.2" data-path="ttest.html"><a href="ttest.html#data-exploration"><i class="fa fa-check"></i><b>2.4.2</b> Data exploration</a></li>
<li class="chapter" data-level="2.4.3" data-path="ttest.html"><a href="ttest.html#the-t-test"><i class="fa fa-check"></i><b>2.4.3</b> The t-test</a></li>
<li class="chapter" data-level="2.4.4" data-path="ttest.html"><a href="ttest.html#anova"><i class="fa fa-check"></i><b>2.4.4</b> ANOVA</a></li>
</ul></li>
<li class="chapter" data-level="2.5" data-path="lines-helpers-regression-lines-loess-and-density-lines-curves.html"><a href="lines-helpers-regression-lines-loess-and-density-lines-curves.html"><i class="fa fa-check"></i><b>2.5</b> Lines: Helpers, Regression lines, Loess and Density Lines &amp; Curves</a><ul>
<li class="chapter" data-level="2.5.1" data-path="lines-helpers-regression-lines-loess-and-density-lines-curves.html"><a href="lines-helpers-regression-lines-loess-and-density-lines-curves.html#drawing-curves-from-functions"><i class="fa fa-check"></i><b>2.5.1</b> Drawing curves from functions</a></li>
</ul></li>
</ul></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Data Analysis and Visualization using R</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="embeddeddf" class="section level2">
<h2><span class="header-section-number">2.3</span> Processing Embedded Dataframes</h2>
<p>Suppose you have imported some data that has a structure like this</p>
<div class="sourceCode" id="cb38"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb38-1" data-line-number="1">genes &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;gene A&quot;</span>, <span class="st">&quot;gene B&quot;</span>, <span class="st">&quot;gene C&quot;</span>, <span class="st">&quot;gene D&quot;</span>)</a>
<a class="sourceLine" id="cb38-2" data-line-number="2">positions &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;chr01:128757:129667&quot;</span>, </a>
<a class="sourceLine" id="cb38-3" data-line-number="3">               <span class="st">&quot;chr01:366389:486990&quot;</span>,</a>
<a class="sourceLine" id="cb38-4" data-line-number="4">               <span class="st">&quot;chr02:8986463:9100856&quot;</span>,</a>
<a class="sourceLine" id="cb38-5" data-line-number="5">               <span class="st">&quot;chr03:53536:87201&quot;</span>)</a>
<a class="sourceLine" id="cb38-6" data-line-number="6">my.genome &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">gene =</span> genes, <span class="dt">position =</span> positions)</a>
<a class="sourceLine" id="cb38-7" data-line-number="7">my.genome</a></code></pre></div>
<pre><code>##     gene              position
## 1 gene A   chr01:128757:129667
## 2 gene B   chr01:366389:486990
## 3 gene C chr02:8986463:9100856
## 4 gene D     chr03:53536:87201</code></pre>
<p>The problem here is that the second column, <code>positions</code>, of type <code>character</code>, actually holds three different variables: the chromosome identifyer, the start position and the stop position on the chromosome. To be able to perform analyses of chromosomal contents, or positional contexts, we will need to split this column into separate columns, each holding exactly one variable of the correct type (<code>factor</code>, <code>integer</code> and <code>integer</code>).</p>
<p>When I first encountered this type of problem (it is a <em>challenge</em> actually, some teachers would object, not a <em>problem</em>…), my first thought was “easy, simply apply a split and bind as three columns”.</p>
<p>Let’s have a look at how the <code>strsplit</code> function works in splitting strings</p>
<div class="sourceCode" id="cb40"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb40-1" data-line-number="1"><span class="kw">strsplit</span>(<span class="dt">x =</span> positions[<span class="dv">1</span><span class="op">:</span><span class="dv">2</span>], <span class="dt">split =</span> <span class="st">&quot;:&quot;</span>)</a></code></pre></div>
<pre><code>## [[1]]
## [1] &quot;chr01&quot;  &quot;128757&quot; &quot;129667&quot;
## 
## [[2]]
## [1] &quot;chr01&quot;  &quot;366389&quot; &quot;486990&quot;</code></pre>
<p>As you can see, strsplit generates a list of vectors, with each vector corresponding to the string at the same index of the original character vector.
So, easy, I thought. Simply assign these elements to three new columns of the original dataframe (assuming every split character results in a vector of three). I first created the columns, defined my splitter function and then used apply to get the job done</p>
<div class="sourceCode" id="cb42"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb42-1" data-line-number="1">## create columns</a>
<a class="sourceLine" id="cb42-2" data-line-number="2">my.genome[, <span class="kw">c</span>(<span class="st">&quot;chromosome&quot;</span>, <span class="st">&quot;start&quot;</span>, <span class="st">&quot;stop&quot;</span>)] &lt;-<span class="st"> </span><span class="ot">NA</span></a>
<a class="sourceLine" id="cb42-3" data-line-number="3">## define splitter function</a>
<a class="sourceLine" id="cb42-4" data-line-number="4">loc.splitter &lt;-<span class="st"> </span><span class="cf">function</span>(x) {</a>
<a class="sourceLine" id="cb42-5" data-line-number="5">    ## strsplit returns a list!</a>
<a class="sourceLine" id="cb42-6" data-line-number="6">    <span class="kw">strsplit</span>(x[<span class="st">&quot;position&quot;</span>], <span class="st">&quot;:&quot;</span>)[[<span class="dv">1</span>]]</a>
<a class="sourceLine" id="cb42-7" data-line-number="7">}</a>
<a class="sourceLine" id="cb42-8" data-line-number="8">## use apply to fill the columns</a>
<a class="sourceLine" id="cb42-9" data-line-number="9">my.genome[, <span class="dv">3</span><span class="op">:</span><span class="dv">5</span>] &lt;-<span class="st"> </span><span class="kw">apply</span>(<span class="dt">X =</span> my.genome,</a>
<a class="sourceLine" id="cb42-10" data-line-number="10">                          <span class="dt">MARGIN =</span> <span class="dv">1</span>,</a>
<a class="sourceLine" id="cb42-11" data-line-number="11">                          <span class="dt">FUN =</span> loc.splitter)</a>
<a class="sourceLine" id="cb42-12" data-line-number="12">my.genome</a></code></pre></div>
<pre><code>##     gene              position chromosome   start    stop
## 1 gene A   chr01:128757:129667      chr01  366389 9100856
## 2 gene B   chr01:366389:486990     128757  486990   chr03
## 3 gene C chr02:8986463:9100856     129667   chr02   53536
## 4 gene D     chr03:53536:87201      chr01 8986463   87201</code></pre>
<p>Whoa, what happened here?! This was not what I had in mind. Can you figure out what happened?</p>
<p>…</p>
<p>I did figure it out (eventually…). The applied function returned three elements at a time, and I had apply fill three columns of my dataframe. And that is exactly what R did, fill the three columns, but not by row but by column! Have a look at the output from apply and you can see:</p>
<div class="sourceCode" id="cb44"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb44-1" data-line-number="1"><span class="kw">apply</span>(<span class="dt">X =</span> my.genome,</a>
<a class="sourceLine" id="cb44-2" data-line-number="2">      <span class="dt">MARGIN =</span> <span class="dv">1</span>,</a>
<a class="sourceLine" id="cb44-3" data-line-number="3">      <span class="dt">FUN =</span> loc.splitter)</a></code></pre></div>
<pre><code>##      [,1]     [,2]     [,3]      [,4]   
## [1,] &quot;chr01&quot;  &quot;chr01&quot;  &quot;chr02&quot;   &quot;chr03&quot;
## [2,] &quot;128757&quot; &quot;366389&quot; &quot;8986463&quot; &quot;53536&quot;
## [3,] &quot;129667&quot; &quot;486990&quot; &quot;9100856&quot; &quot;87201&quot;</code></pre>
<p>Fortunately, R has a function to transpose this kind of structure (a matrix actually): the <code>t()</code> function, so that is what I did:</p>
<div class="sourceCode" id="cb46"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb46-1" data-line-number="1">my.genome[, <span class="dv">3</span><span class="op">:</span><span class="dv">5</span>] &lt;-<span class="st"> </span><span class="kw">t</span>(<span class="kw">apply</span>(<span class="dt">X =</span> my.genome,</a>
<a class="sourceLine" id="cb46-2" data-line-number="2">                            <span class="dt">MARGIN =</span> <span class="dv">1</span>,</a>
<a class="sourceLine" id="cb46-3" data-line-number="3">                            <span class="dt">FUN =</span> loc.splitter))</a>
<a class="sourceLine" id="cb46-4" data-line-number="4">my.genome</a></code></pre></div>
<pre><code>##     gene              position chromosome   start    stop
## 1 gene A   chr01:128757:129667      chr01  128757  129667
## 2 gene B   chr01:366389:486990      chr01  366389  486990
## 3 gene C chr02:8986463:9100856      chr02 8986463 9100856
## 4 gene D     chr03:53536:87201      chr03   53536   87201</code></pre>
<p>Yeah, that’s what I’m talking about! (Feeling very happy with myself…until I googled this problem). I found out there are a gazillion solutions to this problem, but only one of them is very very simple, because it uses a function you know really well: <code>read.table</code>, but not with the <code>file =</code> argument but with <code>text =</code>:</p>
<div class="sourceCode" id="cb48"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb48-1" data-line-number="1">my.genome &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">gene =</span> genes, <span class="dt">position =</span> positions)</a>
<a class="sourceLine" id="cb48-2" data-line-number="2">my.genome &lt;-<span class="st"> </span><span class="kw">cbind</span>(</a>
<a class="sourceLine" id="cb48-3" data-line-number="3">    my.genome,</a>
<a class="sourceLine" id="cb48-4" data-line-number="4">    <span class="kw">read.table</span>(</a>
<a class="sourceLine" id="cb48-5" data-line-number="5">        <span class="dt">text =</span> <span class="kw">as.character</span>(my.genome<span class="op">$</span>position),</a>
<a class="sourceLine" id="cb48-6" data-line-number="6">        <span class="dt">sep =</span> <span class="st">&quot;:&quot;</span>))</a>
<a class="sourceLine" id="cb48-7" data-line-number="7"><span class="kw">colnames</span>(my.genome) &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="kw">colnames</span>(my.genome)[<span class="dv">1</span><span class="op">:</span><span class="dv">2</span>], <span class="st">&quot;chr&quot;</span>, <span class="st">&quot;start&quot;</span>, <span class="st">&quot;stop&quot;</span>)</a>
<a class="sourceLine" id="cb48-8" data-line-number="8">my.genome</a></code></pre></div>
<pre><code>##     gene              position   chr   start    stop
## 1 gene A   chr01:128757:129667 chr01  128757  129667
## 2 gene B   chr01:366389:486990 chr01  366389  486990
## 3 gene C chr02:8986463:9100856 chr02 8986463 9100856
## 4 gene D     chr03:53536:87201 chr03   53536   87201</code></pre>
<p>That’s it. The lessons learned here:</p>
<ul>
<li>Always know that GIYF (Google Is Your Friend)</li>
<li>When reading tables, also those embedded within others, use <code>read.table</code></li>
<li>You really learn a lot by fiddling about with data</li>
</ul>
<hr />
</div>
            </section>

          </div>
        </div>
      </div>
<a href="apply.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="ttest.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"google": false,
"linkedin": false,
"weibo": false,
"instapper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/MichielNoback/R_data_analysis_and_visualization/edit/master/course_syllabus/01-use_cases.Rmd",
"text": "Edit"
},
"download": null,
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://cdn.bootcss.com/mathjax/2.7.1/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:" && /^https?:/.test(src))
      src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
